s3/pkg/service/object.go

```diff
@@ -363,6 +363,130 @@ func (s *s3Service) GetObject(ctx context.Context, req *pb.GetObjectInput, strea
 	return nil
 }
 
+func (s *s3Service) UpdateObjectMeta(ctx context.Context, in *pb.Object, out *pb.PutObjectResponse) error {
+	log.Infoln("UpdateObjectMeta is called in s3 service.")
+	var err error
+	defer func() {
+		out.ErrorCode = GetErrCode(err)
+	}()
+
+	err = s.MetaStorage.UpdateObjectMeta(&meta.Object{Object: in})
+	if err != nil {
+		log.Errorf("failed to update object meta storage, err:", err)
+		err = ErrInternalError
+		return err
+	}
+	out.LastModified = in.LastModified
+	out.Md5 = in.Etag
+	out.VersionId = in.GetVersionId()
+
+	return nil
+}
+
+func (s *s3Service) CopyObject(ctx context.Context, in *pb.CopyObjectRequest, out *pb.CopyObjectResponse) error {
+	log.Infoln("CopyObject is called in s3 service.")
+	var err error
+	defer func() {
+		out.ErrorCode = GetErrCode(err)
+	}()
+
+	srcBucketName := in.SrcBucketName
+	srcObjectName := in.SrcObjectName
+	targetBucketName := in.TargetBucketName
+	targetObjectName := in.TargetObjectName
+	srcBucket, err := s.MetaStorage.GetBucket(ctx, srcBucketName, true)
+	if err != nil {
+		log.Errorln("get bucket failed with err:", err)
+		return err
+	}
+	srcObject, err := s.MetaStorage.GetObject(ctx, srcBucketName, srcObjectName, true)
+	if err != nil {
+		log.Errorln("failed to get object info from meta storage. err:", err)
+		return err
+	}
+	backendName := srcBucket.DefaultLocation
+	if srcObject.Location != "" {
+		backendName = srcObject.Location
+	}
+	srcBackend, err := getBackend(ctx, s.backendClient, backendName)
+	if err != nil {
+		log.Errorln("failed to get backend client with err:", err)
+		return err
+	}
+	srcSd, err := driver.CreateStorageDriver(srcBackend.Type, srcBackend)
+	if err != nil {
+		log.Errorln("failed to create storage. err:", err)
+		return err
+	}
+	reader, err := srcSd.Get(ctx, srcObject.Object, 0, srcObject.Size)
```

Comment:
IMHO, it's better to get target bucket and  target backend, and create target storage driver before get data from source backend.

Reply:
thanks for your reviewing, I will do it as you proposed in next commit